File: packages/ui/src/kg/frontmatter.ts
export interface FrontmatterDocument {
  data: Record<string, unknown>;
  body: string;
}

export function parseFrontmatter(src: string): FrontmatterDocument {
  if (!src.startsWith('---')) return { data: {}, body: src };
  const end = src.indexOf('\n---', 3);
  if (end === -1) return { data: {}, body: src };

  const raw = src.slice(3, end).trim();
  const body = src.slice(end).replace(/^\n---\s*\r?\n?/, '');
  return { data: parseFlatYaml(raw), body };
}

function parseFlatYaml(src: string): Record<string, unknown> {
  const data: Record<string, unknown> = {};
  for (const line of src.split(/\r?\n/)) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('#')) continue;
    const colon = trimmed.indexOf(':');
    if (colon === -1) continue;
    const key = trimmed.slice(0, colon).trim();
    const raw = trimmed.slice(colon + 1).trim();
    if (!key) continue;
    data[key] = parseScalar(raw);
  }
  return data;
}

function parseScalar(value: string): unknown {
  if (!value) return '';
  if ((value.startsWith('"') && value.endsWith('"')) || (value.startsWith("'") && value.endsWith("'"))) {
    return value.slice(1, -1);
  }
  if (value === 'true') return true;
  if (value === 'false') return false;
  const numberValue = Number(value);
  if (Number.isFinite(numberValue) && /^-?\d+(\.\d+)?$/.test(value)) return numberValue;
  return value;
}

File: pkg/ask/frontmatter.go
package ask

import (
	"regexp"
	"strconv"
	"strings"
)

type FrontmatterDocument struct {
	Data map[string]any
	Body string
}

func ParseFrontmatter(src string) FrontmatterDocument {
	if !strings.HasPrefix(src, "---") {
		return FrontmatterDocument{Data: map[string]any{}, Body: src}
	}
	end := strings.Index(src[3:], "\n---")
	if end == -1 {
		return FrontmatterDocument{Data: map[string]any{}, Body: src}
	}
	end += 3
	raw := strings.TrimSpace(src[3:end])
	body := src[end:]
	body = regexpFrontmatterEnd.ReplaceAllString(body, "")
	return FrontmatterDocument{Data: parseFlatYAML(raw), Body: body}
}

var (
	regexpFrontmatterEnd = regexp.MustCompile(`^\n---\s*\r?\n?`)
	regexpNumber         = regexp.MustCompile(`^-?\d+(\.\d+)?$`)
)

func parseFlatYAML(src string) map[string]any {
	data := map[string]any{}
	for _, line := range splitLines(src) {
		trimmed := strings.TrimSpace(line)
		if trimmed == "" || strings.HasPrefix(trimmed, "#") {
			continue
		}
		colon := strings.Index(trimmed, ":")
		if colon == -1 {
			continue
		}
		key := strings.TrimSpace(trimmed[:colon])
		raw := strings.TrimSpace(trimmed[colon+1:])
		if key == "" {
			continue
		}
		data[key] = parseScalar(raw)
	}
	return data
}

func parseScalar(value string) any {
	if value == "" {
		return ""
	}
	if len(value) >= 2 {
		if (strings.HasPrefix(value, `"`) && strings.HasSuffix(value, `"`)) ||
			(strings.HasPrefix(value, `'`) && strings.HasSuffix(value, `'`)) {
			return value[1 : len(value)-1]
		}
	}
	if value == "true" {
		return true
	}
	if value == "false" {
		return false
	}
	if regexpNumber.MatchString(value) {
		if number, err := strconv.ParseFloat(value, 64); err == nil {
			return number
		}
	}
	return value
}
